### PR TITLE
Set options.limit to 1 for non-global string matching

### PR DESCRIPTION
**Pull request by bryce-gibson**
When doing string matching and using regExpOptions without the global flag, it's
possible to get in an infinite loop. Use the same check as matchFromRegex to
avoid that.

Was mucking around with some replacing, and found I was in an infinite loop because of this, so thought it might be helpful? :-)

### CODE REVIEW

**Comment by mehtaphysical** [comment]
thanks! good catch :).

I wonder what the best way to handle this is. I think we should defer to options.limit if a string (not a regex) is passed in. so maybe

```
if (options.limit > 1 && options.regExpOptions.indexOf('g') < 0) {
    options.regExpOptions += 'g'
}
```

Or we should prevent regex options from being passed if we are doing a string replace. So change line 15:

```
 if (!isRegex)
    options.regExpOptions = 'gm';
```

what are your thoughts @bryce-gibson?

**Comment by bryce-gibson** [comment]
I kinda think the second is the way to go (preventing regExpOptions being passed for non regexes).

It seems a bit strange, and hence a bit of a trap, that doing a string replacement can be affected by the regExpOptions provided.

It's less backwards compatible, which is a bit of a concern, but it feels like the more "correct" way for it to operate.

A quick side thought might be to do

    if (!isRegex)
        options.regExpOptions = options.ignoreCase ? 'gmi' : 'gm'

Although, that might be a little more confusing?


My general inkling is that the string replacement syntax should be rigid and have options/flags for specific functionality, whilst the regex version is more "do what you want".

**Comment by bryce-gibson** [comment]
Hmm, I'm now wondering...

Could/should `options.regExpOptions` be removed all together?

And have

    function matchFromString(s, options) {
      return new RegExp(escapeRegExp(s), options.ignoreCase ? 'gmi' : 'gm');
    }

and

    function matchFromRegex(regex, options) {
        if (!regex.global) { // This definitely works in node (I've tried 0.10 latest and 0.12 latest)
          options.limit = 1;
        }
        return regex; // The `new RegExp` may be important for a reason I haven't considered...
    }


Just a thought :-)

**Comment by mehtaphysical** [comment]
i agree with your inkling. Make string replacement rigid with some flags and regex flexible.

I think your idea is solid. Although there was a previous pr to add the regExpOptions flags (https://github.com/eugeneware/replacestream/pull/2). Since that was before regex was an option i think it is fine removing it.

You want to re-write this per your latest comment?

**Comment by bryce-gibson** [comment]
I think it's more in line with that idea now?

This has removed the `regExpOptions` flag and added a new `ignoreCase` option, which defaults to true (effectively).

Happy to tweak as needed :-)

File: index.js [hunk 1]
```diff
@@ -125,19 +122,14 @@ function escapeRegExp(s) {
...
-  } else {
-    var flags = s.toString().replace(/\/[^\/].*\//, '')
-    // If there is no global flag then there can only be one match
-    if (flags.indexOf('g') < 0) {
-      options.limit = 1;
-    }
-    return new RegExp(s.source, flags)
+function matchFromRegex(regex, options) {
+  // If there is no global flag then there can only be one match
+  if (!regex.global) {
```
Comment (mehtaphysical) [inline index.js]: here you should check if the global flag is within the regex itself. you can get the flags like this:

```
var flags = s.toString().replace(/\/[^\/].*\//, '')
```
then just check if there is a g.

**Comment by mehtaphysical** [comment]
looking good. just the one in-line comment

File: index.js (hunk 1 above)
Comment (bryce-gibson) [inline index.js]: The `global` is a property of a RegExp indicating if the global flag is set (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/global).

Hence, I believe it's a simpler way of achieving the same thing?

It looks to be supported in node v0.10 and v0.12...

Comment (mehtaphysical) [inline index.js]: oh nice. my bad. looks good then. for some reason i thought this said options.global. Either way i didnt know about that property :)

**Comment by mehtaphysical** [comment]
looks good!

so im thinking now (sorry) that maybe we should leave the regExpOptions there, while keeping your changes. We'll add your changes to the docs (ignoreCase). And maybe a note that regExpOptions are deprecated.

Also, we can add a comment above regExpOptions that we can remove it upon next major release.

I'm not sure we want to break backwards compatibility. then we can bump this to 2.1.0 instead of 3.0.0.

Thoughts?

**Comment by bryce-gibson** [comment]
Yeah, good plan :-)

I'll give it a look tomorrow, if that works :-)

**Comment by bryce-gibson** [comment]
Done, I think?

Readded the option, marked the tests which use it as deprecated (seemed a good idea?), and added some (very very basic) doc.

**Comment by mehtaphysical** [comment]
looks good! way cleaner than before too :). i'll take one more pass over it tomorrow and then merge. 

thanks again

**Comment by mehtaphysical** [comment]
thanks again. everything is merged and published on npm.

We should add you to the list of contributors :). should I put your name down as Bryce Gibson?

**Comment by bryce-gibson** [comment]
Oh, cool :-) Yep, Bryce Gibson :-)

**Comment by eugeneware** [comment]
Thanks @bryce-gibson and @mehtaphysical - been in hospital for the past month. Thanks for the PR and review guys!

**Comment by mehtaphysical** [comment]
@eugeneware I hope everything is ok :)

**Pull request merged**